### ISSUE TITLE
Fetch External Artifact task gives generic Unprocessable Entity error for failing validations

### ISSUE THREAD

**Issue by chrisb2244** [gocd/gocd#10668]
##### Issue Type
- Bug Report

##### Summary
Validation failures for ArtifactStoreConfig entries produce a nice GUI representation and indicate the location and error message for the failure(s).
Validation failures for FetchArtifactConfig entries produce unprocessable entity responses and a less pleasing GUI/feedback.

Although I encountered this whilst writing a custom plugin, the same can be seen using the [Docker artifact plugin](https://github.com/gocd/docker-registry-artifact-plugin).

Further discussion and some troubleshooting can be found at discussion https://github.com/gocd/gocd/discussions/10639

###### Basic environment details
* Go Version: `22.1.0`
* JAVA Version: `17.0.2`
* OS: `Windows 10`
* Browser vendor and version (if relevant): `Chrome 103.0`

##### Steps to Reproduce
See https://github.com/gocd/gocd/discussions/10639 for details

1. Download and install the [Docker artifact plugin]((https://github.com/gocd/docker-registry-artifact-plugin). (releases page, `1.3.1-270` was the version tested)
2. Create a valid(atable) Docker artifact store (no need to have a working store, junk values for `RegistryURL`, `Username` and `Password` are sufficient when choosing `other` for `RegistryType`)
3. Create a new pipeline with two stages
4. In the first stage, create an External artifact for the store created in step 2, provide some value for `id` and `Image` (test-docker-image and 'gocd/image' chosen here)
5. In the second stage, create a FetchArtifact task, choose `External` for type and then the first stage for the source and cd.go.artifact.docker.registry for the Plugin Id (might get a warning about not being able to determine the plugin, I think this happens because not built previously?)
6. (Optionally) Click save - configuration is valid and saves.
7. Set the `Environment Variable Prefix` to "1" (validation requires it not start with a number)
8. Click save - observe the error shown in Actual Results section

##### Expected Results
Feedback similar to this (this is for an invalid artifact store config, which displays well):
![image](https://user-images.githubusercontent.com/7201433/181703215-ad89e661-78d5-46ca-bfc2-ff2b7bee3d3a.png)

##### Actual Results
![image](https://user-images.githubusercontent.com/7201433/183010885-fd8b3055-5a2d-45a2-97be-ffa31c952278.png)

##### Possible Fix
Unknown

##### Log snippets
Some logs are given in #10639

##### Code snippets/Screenshots
Some code snippets are given in #10639 but mainly indicate that the output of the validation request is (as far as I can tell) of the structure specified by https://plugin-api.gocd.org/current/artifacts/#validate-fetch-artifact-config

**Comment by chadlwilson** [comment]
I think I have a fix for this in #10679. I'll let you know when a test build is ready.

There's also something funny with the Plugin Id setting if you don't set the Pipeline (which is actually optional if you want it to fetch from current pipeline). Did you notice that? (it's cropped off, but red in your screenshot above) If you enter some text and then remove it from Pipeline the validation fixes itself. Sigh :-)

![image](https://user-images.githubusercontent.com/29788154/183463613-5a3410a3-3481-450f-9aa6-33fe6fdf44ea.png)

**Comment by chrisb2244** [comment]
I had noticed that sometimes the plugin detection complained and sometimes it didn't, but I hadn't worked out what triggered it.

I had thought it was related to having built the pipeline (or maybe more accurately the pipeline/stage that produces the fetched artifact) but I guess not.

I think in all the cases I tested, I did enter a pipeline value, but I noticed it was described as optional. I think if you don't set it, maybe you don't get hints for the later fields? (I don't mind this, just that was I think why I set it). In some cases I was testing fetching from a different pipeline, so then I had to set it.

**Comment by chadlwilson** [comment]
Thanks for the stunning bug report, made it very easy to replicate and track it down👍

**Comment by chadlwilson** [comment]
In effect this is similar to #2553 and #7423

**Comment by chadlwilson** [comment]
@chrisb2244 if you'd like to give it a spin with your custom plugin there are builds available at [windows server installer](https://nexus.gocd.io/repository/s3-mirrors/local/other/go-server-22.2.0-3-jre-64bit-setup.exe) and [generic zip](https://nexus.gocd.io/repository/s3-mirrors/local/other/go-server-22.2.0-3.zip).

If you are using the installer, please don't do so on the same host as your "main" installation (for some reason Windows installer blocks downgrades, even though generally GoCD downgrades are fine) or anything you cannot afford to lose, in case there is something awry. Should be fine, as has gone through the full regression suite and am intending to cut an official 22.2.0 release shortly, but ya, best to be conservative.

**Comment by chrisb2244** [comment]
A new task produced the same behaviour - I think the error message appearing twice is intended behaviour (since two fields have the error, and I guess the top shows a concatenated list?) but not sure why there's still a "There was an unknown error..." message :/

**Comment by chadlwilson** [comment]
Thanks for the feedback. At least this is an improvement (the cause is not lost) and I assume it unblocks you.

Good catch on the top error. This seems to be a related, but potentially separate issue - it might take me a bit longer to figure out the intention. I think there are potentially 4 issues:

1. The "overall" error should be displayed as `Validations failed for pipeline 'blah'. Error(s): [Validation failed.]. Please correct and resubmit.` rather than `...unknown error...`. I can see why this is happening and believe it's just a logic error. I have it changed locally, but will see whether it's safe to change and do so if possible.
2. By rights, the additional messages relating to the individual fields at the top are supposed to not display again if they are rendered in the below dialog. The intent is the the UI knows which errors have been "consumed" by a field, and only displays unconsumed ones, to ensure they are not lost. I don't know why this isn't working here, and whether this is something specific to plugins and/or artifact plugins. Will definitely require more digging.
3. _Nitpick_: When concatenating messages for display, they should be separated by a space at least.
4. _Nitpick_: Messages could be de-duplicated for display.

Does that make sense?

**Comment by chadlwilson** [comment]
Updated [windows server installer](https://nexus.gocd.io/repository/s3-mirrors/local/other/go-server-22.2.0-11-jre-64bit-setup.exe) and [generic zip](https://nexus.gocd.io/repository/s3-mirrors/local/other/go-server-22.2.0-11.zip).

**Comment by chrisb2244** [comment]
LGTM - I get the validations failed message now (no more unknown error), and only a single instance of the message at the top.

If according to your point 2 I shouldn't see the message at all, if it appears lower down, I guess this might be something I've done wrong in "consuming" the error.
I'll check with the Docker plugin and see how it looks there.

**Comment by chadlwilson** [comment]
For point 2, I think it is either a bug or a design limitation with how the team were able to implement the pipeline config and the modals specifically for task config. The docker plugin has the same issue, as with editing other plugin task config for custom tasks.

The challenge, I believe, is that the API accepts the whole pipeline config, but the modal is actually only allowing you to edit a subset of it.

 From debugging the code I can see why it's not able to avoid repeating any discovered validation errors but I couldn't quite figure out how it's transferring the errors to the plugin such that the plugin view can render them. Other parts of pipeline config take a slightly different approach to determining errors already consumed by specific fields/view components.

 I'll create a separate issue and seek some of the earlier team input specifically I think. I would assume it's not something you're doing wrong here.

**Comment by chadlwilson** [comment]
Created the separate issue at #10685.

The changes here are merged and proper code-signed experimental builds are available via https://www.gocd.org/download/#windows

Looking to cut a release of `22.2.0` shortly, so if you notice something seriously broken, speak now for forever hold your peace 😅
